concordium-consensus/src/Concordium/GlobalState/PurgeTransactions.hs

```diff
@@ -107,20 +107,35 @@ purgeTables lastFinCommitPoint oldestArrivalTime currentTime TransactionTable{..
                 | otherwise = (mmnonce <> Just (Max n), Just ts')
         put (mmnonce', tht')
         return mres
-    -- Purge the non-finalized transactions for a specific account.
-    purgeAccount :: AccountAddressEq -> AccountNonFinalizedTransactions -> State (PendingTransactionTable, TransactionHashTable) AccountNonFinalizedTransactions
-    purgeAccount addr AccountNonFinalizedTransactions{..} = do
-        (ptt0, trs0) <- get
+    -- Purge the non-finalized transactions for an account,
+    -- accumulating the non-finalized transactions that were
+    -- not purged. Purging the transactions removes them
+    -- from the transaction table and updates the pending
+    -- transaction table accordingly.
+    purgeAccount ::
+        -- accumulator
+        ( HM.HashMap AccountAddressEq AccountNonFinalizedTransactions,
+          (PendingTransactionTable, TransactionHashTable)
+        ) ->
+        -- current entry key
+        AccountAddressEq ->
+        -- current entry value
+        AccountNonFinalizedTransactions ->
+        -- result
+        ( HM.HashMap AccountAddressEq AccountNonFinalizedTransactions,
+          (PendingTransactionTable, TransactionHashTable)
+        )
+    purgeAccount (anfts, (ptt0, trs0)) addr AccountNonFinalizedTransactions{..} =
         -- Purge the transactions from the transaction table.
         let (newANFTMap, (mmax, !trs1)) = runState (Map.traverseMaybeWithKey purgeTxs _anftMap) (Nothing, trs0)
-        -- Update the pending transaction table.
-        let updptt (Just (Max newHigh)) (Just (low, _))
+            -- Update the pending transaction table.
+            updptt (Just (Max newHigh)) (Just (low, _))
                 | newHigh < low = Nothing
                 | otherwise = Just (low, newHigh)
             updptt _ _ = Nothing
             !ptt1 = ptt0 & pttWithSender . at' addr %~ updptt mmax
-        put (ptt1, trs1)
-        return AccountNonFinalizedTransactions{_anftMap = newANFTMap, ..}
+            anfts' = if null newANFTMap then anfts else HM.insert addr AccountNonFinalizedTransactions{_anftMap = newANFTMap, ..} anfts
```

Comment:
**Question:** Is this right?

I might be off here, but doesn't `newANFTMap` represent the remaining non-finalized transactions for an account? If so, then when this is empty, we don't update the map, which seems wrong to me.

Reply:
What do you mean by updating the map? If it's empty then we do not add it to the accummulated `ttNonFinalizedTransactions` Map. If there are still non-finalized transactions, then we add it to the accummulated `ttNonFinalizedTransactions`.

Reply:
ahh, my bad then.
I got confused, thought `anfts` was the current entry value (`AccountNonFinalizedTransactions`) but it was actually the accumulator.

Reply:
ah. Yeah.